jep/401/README.adoc

```diff
@@ -0,0 +1,246 @@
+
+This proposal provides an approach to make it extensible to provide a customization mechanism for a better integration that also reduces current tech debt.
+
+== Specification
+
+The main aspect of the change is the introduction of a new extension point `Header` as an abstract class that provides capabilities to render a specific header and a default implementation of that, named `JenkinsHeader` that is enabled by default always.
+
+All headers will provide an implementation of a priority method, via https://javadoc.jenkins.io/hudson/Extension.html[ordinal] though the `Extension` annotation, that will help to override, based on its value, which header will be rendered. Default one will provide a `Integer.MIN_VALUE` value.
+
+On `pageHeader.jelly` file we will perform a refactor to make it more modular providing different sections:
+
+* `headerDocumentation.jelly`: that will provide the content of the documentation of the Jelly page
+* `preHeader.jelly`: that will provide the content of the elements that want to be provided/rendered before header div is rendered on the browser
+* `headerContent.jelly`: that will provide the content of the header div
+* `postHeader.jelly`: that will provide the content of the elements that want to be provided/rendered after the header div is rendered on the browser
+
+Thus, `pageHeader.jelly` should look like:
+
+```
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
+	<j:set var="header" value="${h.header()}" />
+	<st:include it="${header}" page="headerDocumentation.jelly"/>
+	<st:include it="${header}" page="preHeader.jelly"/>
+	<st:include it="${header}" page="headerContent.jelly"/>
+	<st:include it="${header}" page="postHeader.jelly"/>
+</j:jelly>
+```
+
+The `h.header()` method is provided in the `Functions` and it will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value). In case two or more headers are provided via external plugins with the same priority (and max), then the default one will be provided and a warning message would be provided on the logs.
+
+In terms of simplicity, proposal aims to use only one extension to do a full replacement of the header. As a follow up, each mentioned section of the header could have its own extension to override only a specific part of the header.
+
+=== Extensibility
+
+Jenkins users will be able to interact with the `Header` extension as follows in case want to customize the header of their instances via plugin:
+
+* Extend `Header` and provide an implementation of the priority method giving to this header the max possible value (via ordinal value)
+* Use an `ExtensionFilter` that would make it impossible to provide any additional Header instance
```

Comment:
Beware of that: You're telling our customers how to get rid of our own header, which goes against any try to protect our own brand. I wouldn't mention the `ExtensionFilter` and let others to figure out how to do it. 
PS: trying to mention Runxia to open the discussion, but I cannot find her

Reply:
I agree with @fcojfernandez

Reply:
Got it, I will remove this part

Reply:
Removed in https://github.com/imonteroperez/jep/pull/1/commits/b240f8f85965918e0fd775ecfad4fe2121fe555b